### PR TITLE
Consistent container update / removal 

### PR DESCRIPTION
**Pull request by etpinard**
fixes https://github.com/plotly/plotly.js/issues/1069

Currently, `relayout` is allowed to be called with :diamond_shape_with_a_dot_inside: special :diamond_shape_with_a_dot_inside: `'add'` and `'remove'` values for annotations and shapes. For example, 

```js
Plotly.relayout(gd, 'annotations', 'add');
// => adds a blank annotations

Plotly.relayout(gd, 'annotations[2]', 'remove');
// => remove the 3rd annotation
```

which gets the job done, but is a little sloppy. We can do better and luckily `'add'` and `'remove'` we're never implemented for the new (e.g. images, updatemenus, sliders, mapbox layers) layout array containers.

This PR enforces,

```js
Plotly.relayout(gd, `${containerName}[${itemIndex}]`, null);
```

as a way to delete items in containers found in layout. Similarly, 

```js
Plotly.restyle(gd, `transforms[${itemIndex}]`, null, [`${traceIndex}`]);
```

will delete transform item in the relevant trace(s).

Moreover, 

```js
Plotly.relayout(gd, `${containerName}[${newItemIndex}]`, {
  // container item opts 
});
```

will consistently add items to the array container if `newItemIndex` is beyond the current container length.

### CODE REVIEW

**Comment by etpinard** [comment]
I made a PR right now even this is a WIP as this was the last remaning item in the [`v1.19.0`](https://github.com/plotly/plotly.js/milestone/12) without a corresponding PR. I'll wrap this is up tomorrow.

TODO:

- [ ] add mucho tests
- [ ] add `restyle` logic for `transforms`

File: src/plots/plots.js [hunk 1]
```diff
@@ -1446,13 +1449,7 @@ plots.extendTrace = function(destTrace, srcTrace) {
...
  */
 plots.extendLayout = function(destLayout, srcLayout) {
-    return plots.extendObjectWithContainers(destLayout, srcLayout, [
-        'annotations',
-        'shapes',
-        'images',
-        'sliders',
-        'updatemenus'
-    ]);
+    return plots.extendObjectWithContainers(destLayout, srcLayout, plots.layoutArrayContainers);
```
Comment (rreusser) [inline src/plots/plots.js]: Love it. Nice refactor.

Comment (etpinard) [inline src/plots/plots.js]: Yep. I thought exposing `plots.layoutArrayContainers` was a nice compromise.

Next step would be to auto-generate that list using the plot-schema. (later ...)

**Comment by rreusser** [comment]
Outstanding! What happens to the surrounding indices when you do that? Presumably it's up to the user/front-end to manage references to non-existent array elements. Like if you have `[annotation1, annotation2]` and you remove the first annotation. Do you then have `[null, annotation2]` or `[annotation2]`?

**Comment by etpinard** [comment]
> Do you then have [null, annotation2] or [annotation2]?

you have `[annotation2]`

That is, `null` clears items in the array containers.

**Comment by bpostlethwaite** [comment]
Is there a place we can add a deprecation notice about the 'add' and 'remove' API? Do any docs need updating? Should we add an issue in the Plotlyjs 2.0 to remove 'add' and 'remove' as part of that Milestone?

**Comment by bpostlethwaite** [comment]
:dancer: for the PR. Just a few questions

**Comment by etpinard** [comment]
> Is there a place we can add a deprecation notice about the 'add' and 'remove' API? Do any docs need updating?

I don't any reference of `'add'` or `'remove'` under https://plot.ly/javascript/#chart-events maybe @cldougl would know more?

> Should we add an issue in the Plotlyjs 2.0 to remove 'add' and 'remove' as part of that Milestone?

Good call. I'll add it to https://github.com/plotly/plotly.js/issues/420

**Comment by rreusser** [comment]
@etpinard That makes sense. I'll only point out the obvious *possibility* to write code that reference other items by array index.

**Pull request merged**